### PR TITLE
Add client Python code

### PR DESCRIPTION
**Pull request by thatch**
It has TODOs and needs tests/testing both, but it's in a state ready to review (license/style both reasonable).  @precision can you take a look?

### CODE REVIEW

File: software/authbox/api.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+import ConfigParser
+import os.path
+import Queue
+import re
+import threading
+
+try:
+  from RPi import GPIO
+  # TODO check version and output error
```
Comment (precision) [inline software/authbox/api.py:29]: You can just check GPIO.VERSION

**Review by precision — APPROVED** [review]

File: software/authbox/api.py (hunk 1 above)
Comment (thatch) [reply software/authbox/api.py:29]: Do you know what version is api-incompatible with what's available now?

Comment (precision) [reply software/authbox/api.py:29]: I think < "0.6"

File: software/authbox/timer.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      self.set_queue.put_nowait(delay)
+    except Queue.Full:
+      raise Exception("There is already a queued timeout")
+
+  def cancel(self):
+    with self.cancel_condition:
+      self.cancel_condition.notify()
+    try:
+      while True:
+        self.set_queue.get_nowait()
```
Comment (gmlewis) [inline software/authbox/timer.py:53]: Is this a full CPU busy-wait (judging from the "get_nowait")?
If so, you might want to add a small delay here to not waste power or heat up the CPU... (or release the CPU to other parts of the system?)
Your call.

File: software/authbox/badgereader_hid_keystroking.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    """
+
+    rfid = ''
+    capitalized = 0
+    device = self.f
+    for event in device.read_loop():
+      data = evdev.categorize(event)
+      if event.type == evdev.ecodes.EV_KEY and data.keystate == 1:
+      # detects if the keyboard uses "LSHFT"
+        if data.scancode == 42:
```
Comment (gmlewis) [inline software/authbox/badgereader_hid_keystroking.py:94]: nit: I would make these 42's constants at the top of the file with descriptive names... I think it is easier to read the code where it is used... your call.
(Here and line 98 below.)

File: software/authbox/badgereader_hid_keystroking.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    Returns:
+      The file for input events that read_input can listen to
+    """
+
+    device_name = '<USBBadgeScanner>'
+    devices = [evdev.InputDevice(x) for x in evdev.list_devices()]
+    device = None
+    for dev in devices:
+      if str(dev.name) == self._device_name:
+        device = dev
```
Comment (gmlewis) [inline software/authbox/badgereader_hid_keystroking.py:72]: Why not just `return dev` here and nix line 69 above?

File: software/authbox/api.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    # N.b. args are from config, kwargs are passed from python
+    options = self.config.get('pins', name).split(':')
+    cls_name = options[0]
+    for c in KNOWN_CLASSES:
+      if c.endswith('.' + cls_name):
+        cls = c
+        break
+    else:
+      raise Exception('Unknown item', name)
+    obj = cls(self.event_queue, *options[1:], **kwargs)
```
Comment (gmlewis) [inline software/authbox/api.py:151]: I'm concerned here about the case where there *are* `KNOWN_CLASSES` but non of them end with `cls_name`...

In that case, no exception is raised, but `cls` is never defined... so line 151 is going to throw an `undefined` exception.

**Review by gmlewis — COMMENTED** [review]
This looks really cool, Tim!  Nice job.  Just a few comments.

**Comment by thatch** [comment]
Thanks for your comments, Glenn.  I sent a new PR https://github.com/google/makerspace-auth/pull/4 that addresses all of these except RPi.GPIO version check.

**Pull request closed without merging**